### ISSUE TITLE
transmute_undefined_repr: false positive on transmutes involving MaybeUninit

### ISSUE THREAD

**Issue by dtolnay** [rust-lang/rust-clippy#8498]
### Summary

MaybeUninit\<T\> has the same layout as T.

### Lint Name

transmute_undefined_repr

### Reproducer

```rust
use std::mem::{self, MaybeUninit};

struct Struct {
    a: u32,
    b: u32,
}

fn f(s: Struct) -> MaybeUninit<Struct> {
    unsafe { mem::transmute::<Struct, MaybeUninit<Struct>>(s) }
}
```

```console
error: transmute from `Struct` which has an undefined layout
 --> src/main.rs:9:14
  |
9 |     unsafe { mem::transmute::<Struct, MaybeUninit<Struct>>(s) }
  |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[deny(clippy::transmute_undefined_repr)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#transmute_undefined_repr
```

> transmute from \`Struct\` which has an undefined layout

Whether or not `Struct` has a defined layout shouldn't be relevant, since MaybeUninit\<T\> has the same layout as T regardless of the layout of T. This is similar to what happened in https://github.com/rust-lang/rust-clippy/issues/8417.

### Version

```text
rustc 1.61.0-nightly (f0c4da499 2022-03-01)
binary: rustc
commit-hash: f0c4da49983aa699f715caf681e3154b445fb60b
commit-date: 2022-03-01
host: x86_64-unknown-linux-gnu
release: 1.61.0-nightly
LLVM version: 14.0.0
```

**Comment by dtolnay** [comment]
More realistic example of this in https://github.com/dtolnay/cxx/blob/c53e4bc81b876c3d19652e9e35d85379239cdc31/src/rust_string.rs.
